Mathlib/RingTheory/PowerSeries/Derivative.lean

```diff
@@ -0,0 +1,176 @@
+  rw [fDerivativeFun, coeff_mk]
+
+theorem fDerivativeFun_coe (f : R[X]) : (f : R⟦X⟧).fDerivativeFun = derivative f := by
+  ext
+  rw [coeff_fDerivativeFun, coeff_coe, coeff_coe, coeff_derivative]
+
+theorem fDerivativeFun_add (f g : R⟦X⟧) :
+    fDerivativeFun (f + g) = fDerivativeFun f + fDerivativeFun g := by
+  ext
+  rw [coeff_fDerivativeFun, map_add, map_add, coeff_fDerivativeFun,
+    coeff_fDerivativeFun, add_mul]
+
+theorem fDerivativeFun_C (r : R) : fDerivativeFun (C R r) = 0 := by
+  ext n
+  rw [coeff_fDerivativeFun, coeff_C]
+  split_ifs with h
+  · cases succ_ne_zero n h
+  · rw [zero_mul, map_zero]
+
+theorem trunc_fDerivativeFun (f : R⟦X⟧) (n : ℕ) :
+    trunc n f.fDerivativeFun = derivative (trunc (n + 1) f):= by
+  ext d
+  rw [coeff_trunc]
+  split_ifs with h
+  · have : d + 1 < n + 1 := succ_lt_succ_iff.2 h
+    rw [coeff_fDerivativeFun, coeff_derivative, coeff_trunc, if_pos this]
+  · have : ¬d + 1 < n + 1 := by rwa [succ_lt_succ_iff]
+    rw [coeff_derivative, coeff_trunc, if_neg this, zero_mul]
+
+--A special case of `fDerivativeFun_mul`, used in its proof.
+private theorem fDerivativeFun_coe_mul_coe (f g : R[X]) : fDerivativeFun (f * g : R⟦X⟧) =
+    f * derivative g + g * derivative f  := by
+  rw [←coe_mul, fDerivativeFun_coe, derivative_mul,
+    add_comm, mul_comm _ g, ←coe_mul, ←coe_mul, Polynomial.coe_add]
+
+/-- Leibniz rule for formal power series.-/
+theorem fDerivativeFun_mul (f g : R⟦X⟧) :
+    fDerivativeFun (f * g) = f • g.fDerivativeFun + g • f.fDerivativeFun := by
+  ext n
+  have h₁ : n < n + 1 := lt_succ_self n
+  have h₂ : n < n + 1 + 1 := Nat.lt_add_right _ _ _ h₁
+  rw [coeff_fDerivativeFun, map_add, coeff_mul_eq_coeff_trunc_mul_trunc _ _ (lt_succ_self _),
+    smul_eq_mul, smul_eq_mul, coeff_mul_eq_coeff_trunc_mul_trunc₂ g f.fDerivativeFun h₂ h₁,
+    coeff_mul_eq_coeff_trunc_mul_trunc₂ f g.fDerivativeFun h₂ h₁, trunc_fDerivativeFun,
+    trunc_fDerivativeFun, ←map_add, ←fDerivativeFun_coe_mul_coe, coeff_fDerivativeFun]
+
+theorem fDerivativeFun_one : fDerivativeFun (1 : R⟦X⟧) = 0 := by
+  rw [←map_one (C R), fDerivativeFun_C (1 : R)]
+
+theorem fDerivativeFun_smul (r : R) (f : R⟦X⟧) : fDerivativeFun (r • f) = r • fDerivativeFun f := by
+  rw [smul_eq_C_mul, smul_eq_C_mul, fDerivativeFun_mul, fDerivativeFun_C, smul_zero, add_zero,
+    smul_eq_mul]
+
+variable (R)
+/--The formal derivative of a formal power series.-/
+noncomputable def fDerivative : Derivation R R⟦X⟧ R⟦X⟧
+where
+  toFun             := fDerivativeFun
+  map_add'          := fDerivativeFun_add
+  map_smul'         := fDerivativeFun_smul
+  map_one_eq_zero'  := fDerivativeFun_one
+  leibniz'          := fDerivativeFun_mul
+/--Abbreviation of `PowerSeries.fDerivative`, the formal derivative on `R⟦X⟧`.-/
+scoped notation "d⁄dX" => fDerivative
+
+variable {R}
+
+@[simp] theorem fDerivative_C (r : R) : d⁄dX R (C R r) = 0 := fDerivativeFun_C r
+
+@[simp] theorem coeff_fDerivative (f : R⟦X⟧) (n : ℕ) :
```

Comment:
Are you sure we want this as a `simp` lemma?

Reply:
Thanks, you're right of course. I'm also making the name change that you suggest (i.e. removing the f).